File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    let maxCount = 0;
    const visited = Array(dungeons.length).fill(false);
    
    function backtrack(fatigue, count) {
        maxCount = Math.max(maxCount, count);
        
        for (let i = 0; i < dungeons.length ; i++) {
            
            if (!visited[i] && fatigue >= dungeons[i][0]) {
                visited[i] = true;
                backtrack(fatigue-dungeons[i][1], count+1);
                visited[i] = false;
            }
            
        }
    }
    
    backtrack(k, 0);
    
    return maxCount;
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    
    visited = [False for _ in range(len(dungeons))]
    
    return backtrack(k, dungeons, visited)

def backtrack(remain, dungeons, visited):
    
    count = sum(visited)
    
    for i, (requirement, consumption) in enumerate(dungeons):

        if not visited[i] and remain >= requirement:

            visited[i] = True
            
            count = max(count, backtrack(remain - consumption, dungeons, visited))
            
            visited[i] = False
    
    return count